src/webgpu/web_platform/copyToTexture/ImageBitmap.spec.ts

```diff
@@ -179,15 +181,79 @@ got [${failedByteActualValues.join(', ')}]`;
 
     return entry.get(color)!;
   }
+
+  // Helper functions to generate imagePixels based input configs.
+  getImagePixels(
+    format: UncompressedTextureFormat,
+    width: number,
+    height: number,
+    transparentOp: TransparentOp,
+    orientationOp: OrientationOp
+  ): Uint8ClampedArray {
+    const bytesPerPixel = kUncompressedTextureFormatInfo[format].bytesPerBlock;
+    if (bytesPerPixel === undefined) {
+      return new Uint8ClampedArray(0);
+    }
+
+    // Generate input contents by iterating 'Color' enum
+    const imagePixels = new Uint8ClampedArray(bytesPerPixel * width * height);
+    const flipYPixels = new Uint8ClampedArray(bytesPerPixel * width * height);
+    const startPixel = Color.Red;
+    for (let i = 0, currentPixel = startPixel; i < width * height; ++i) {
+      const pixelData = this.generatePixel(currentPixel, format);
+      for (let j = 0; j < bytesPerPixel; ++j) {
+        // All pixels are 0 due to premultiply alpha
+        if (transparentOp === 'premultiply' && currentPixel === Color.TransparentBlack) {
+          imagePixels[i * bytesPerPixel + j] = 0;
+        } else {
+          imagePixels[i * bytesPerPixel + j] = pixelData[j];
+        }
+      }
```

Comment:
Sorry for a comment on old code, but I found this a bit confusing. This loop is looping over bytes in the pixel, though that's confusing because it really just wants to copy a whole pixel value. How about this which would express the intent better:

```suggestion
      if (transparentOp === 'premultiply' && currentPixel === Color.TransparentBlack) {
        // All pixels are 0 due to premultiply alpha
        imagePixels.fill(0, i * bytesPerPixel, (i + 1) * bytesPerPixel);
      } else {
        imagePixels.set(pixelData, i * bytesPerPixel);
      }
```

Depending on how much refactoring you want to do of this existing code, I would even suggest folding the premultiplication into the `generatePixel` helper (where it really should be, _prior_ to pack+encode). Then here you _only_ need `imagePixels.set(...)`.

Reply:
Do you have any suggestions on return an correct channel number and all 0 pixelData?

Reply:
Thanks for the details! Almost refactor this as what your suggested. Except for the map part which need to take premultiply and none-premultiply as different entry.(So the parameter part changes a bit too). PTAL!